### PR TITLE
[TrilinosApplication] Extend `trilinos_space.h` with matrix multiplication and extensive testing

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

Extend `trilinos_space.h` with matrix multiplication and extensive testing

**🆕 Changelog**

- [Adding two norm for matrices](https://github.com/KratosMultiphysics/Kratos/commit/df81c414e3be82057d66bb09002949eb30b3ebac)
- [Adding matrix-matrix multiplication](https://github.com/KratosMultiphysics/Kratos/commit/26e790e33a907b93beb4ac5a6ba3694487ca44a2)
- [Adding new tests](https://github.com/KratosMultiphysics/Kratos/commit/f802faa908ce3dac78cd2c58715deab0021e2c6d)
- [Dot product test](https://github.com/KratosMultiphysics/Kratos/commit/5ce9fde222349ec7b422d6b5f77bc78d9536a91a)
- [Adding Min/Max test](https://github.com/KratosMultiphysics/Kratos/commit/c2472fc49ee805e9be0884965d33c5ef24301637)
- [Adding Norm2 tests](https://github.com/KratosMultiphysics/Kratos/commit/c8fc1e76605b045f3097882080205693a6018b5f)
- [ImproveTrilinos CPP utilities](https://github.com/KratosMultiphysics/Kratos/commit/54af93a196d84b651a915e00eaa7e6ea55e08eef)
- [Move TrilinosCPPUtilities to cpp](https://github.com/KratosMultiphysics/Kratos/commit/44fca1198e04f5873a50e233f36e0a8e8e5b2807)
- [New multiplication test](https://github.com/KratosMultiphysics/Kratos/commit/f19976de3b784246125f834d86565fe65be1fb1a)
- [Adding matrix multiplication test](https://github.com/KratosMultiphysics/Kratos/commit/c44966eb39dad6f1fc611b025e7284c8d438a689)
- [Test TrilinosTransposeMultMatrixVector](https://github.com/KratosMultiphysics/Kratos/commit/047bfd57fe2323f7d04415cbe29a624413e10c67)
- [Adding TrilinosTransposeMultMatrixMatrix test](https://github.com/KratosMultiphysics/Kratos/commit/fd885d6dbb02bc4417355aea28e8dc5319cdeced)
- [Adding BDBtProductOperation and BtDBProductOperation](https://github.com/KratosMultiphysics/Kratos/commit/2c469f8743af9e7809c54d0469f0cf4bf7aa7bf4)
- [Adding TrilinosInplaceMult](https://github.com/KratosMultiphysics/Kratos/commit/aa680c5d0950cbd64c4d97149cee76f160fcb770)
- [Adding several methods tests](https://github.com/KratosMultiphysics/Kratos/commit/2d4ece7a93fa54bb9233342b2b89dd1da53cf54a)

### CODE REVIEW

File: applications/TrilinosApplication/trilinos_space.h [hunk 1]
```diff
@@ -31,6 +31,9 @@
 #include <EpetraExt_VectorIn.h>
 #include <EpetraExt_RowMatrixOut.h>
 #include <EpetraExt_MultiVectorOut.h>
+#include <EpetraExt_MatrixMatrix.h>
+
+// NOTE: Epetra will be replaced with Tpetra in the future, there is an intermediate interface called Xpetra, which uses the interface of Toetra but allows to use Epetra
```
Comment (philbucher) [inline applications/TrilinosApplication/trilinos_space.h]: Not sure if it is worth the effort, is TPetra better?

I would rather work towards our own CSR matrix to drop Trilinos

File: applications/TrilinosApplication/trilinos_space.h [hunk 2]
```diff
@@ -278,41 +294,144 @@ class TrilinosSpace
...
+        Epetra_MpiComm& rComm
+        )
+    {
+        // Create a map
+        const int size = Size1(rB);
+        Epetra_Map Map(size, 0, rComm);
+
+        // Create an Epetra_Matrix
+        std::vector<int> NumNz;
+        MatrixType aux(::Copy, Map, NumNz.data());
```
Comment (philbucher) [inline applications/TrilinosApplication/trilinos_space.h]: It is not possible do without the Copy?

Iirc Aditya said that this causes memory spikes

File: applications/TrilinosApplication/trilinos_space.h [hunk 3]
```diff
@@ -278,41 +294,144 @@ class TrilinosSpace
...
+     * @param rA The first matrix considered
+     * @param rB The second matrix considered
+     * @param rC The result of the multiplication
+     * @param TransposeFlag Flags to transpose the matrices
+     */
+    static void TransposeMult(
+        const MatrixType& rA,
+        const MatrixType& rB,
+        MatrixType& rC,
+        const std::vector<bool> TransposeFlag = {false, false}
```
Comment (philbucher) [inline applications/TrilinosApplication/trilinos_space.h]: Please use std::pair

File: applications/TrilinosApplication/trilinos_space.h [hunk 4]
```diff
@@ -278,41 +294,144 @@ class TrilinosSpace
...
+     * @param rB The second matrix considered
+     * @param rC The result of the multiplication
+     */
+    static void Mult(
+        const MatrixType& rA,
+        const MatrixType& rB,
+        MatrixType& rC
+        )
+    {
+        const bool transpose_flag = false;
```
Comment (philbucher) [inline applications/TrilinosApplication/trilinos_space.h]: Constexpr ;)

**Review by philbucher — CHANGES_REQUESTED** [review]
Nice 👍🏻 

Will review more in detail later

In the meantime you could separate the cosmetic changes and the tests to speed up the review process ;)

File: applications/TrilinosApplication/trilinos_space.h (hunk 1 above)
Comment (loumalouomega) [reply applications/TrilinosApplication/trilinos_space.h]: (Just a comment, but according to the official benchmarks is better...)

File: applications/TrilinosApplication/trilinos_space.h (hunk 2 above)
Comment (loumalouomega) [reply applications/TrilinosApplication/trilinos_space.h]: There are only two options, Copy or View, I will check if it works. In any case this is **an empty matrix**, the memory should be not spiked. If you see old implementations it uses Copy too

File: applications/TrilinosApplication/trilinos_space.h (hunk 3 above)
Comment (loumalouomega) [reply applications/TrilinosApplication/trilinos_space.h]: Okay

File: applications/TrilinosApplication/trilinos_space.h (hunk 4 above)
Comment (loumalouomega) [reply applications/TrilinosApplication/trilinos_space.h]: Okay

Comment (loumalouomega) [reply applications/TrilinosApplication/trilinos_space.h]: Done

File: applications/TrilinosApplication/trilinos_space.h (hunk 3 above)
Comment (loumalouomega) [reply applications/TrilinosApplication/trilinos_space.h]: Done

File: applications/TrilinosApplication/trilinos_space.h (hunk 2 above)
Comment (loumalouomega) [reply applications/TrilinosApplication/trilinos_space.h]: Done

File: applications/TrilinosApplication/trilinos_space.h (hunk 1 above)
Comment (loumalouomega) [reply applications/TrilinosApplication/trilinos_space.h]: Comment removed

**Review by philbucher — APPROVED** [review]
nice

**Pull request merged**